pkg/validation/validation-rule.md

```diff
@@ -0,0 +1,54 @@
+### Id and Name:
+`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
+
+The restriction is added to allow easy translation to K8 resource names, and also to have consistent rules for both name and id fields
+The validation will be done as part of schema validation, the rule will be introduced as a regex in schema definition, any objection of the rule in devfile will result in a failure.
+
+- limit to lowercase characters i.e.; no uppercase allowed
+- limit within 63 characters, except for endpoint’s name which is limited within 15 characters
+- no special characters allowed, only dash(-)
+- start with an alphanumeric character
+- end with an alphanumeric character
+- must not be with all numbers (not included in schema validation, will have extra validation function)
+
+
+### Endpoints:
+- all the endpoint names are unique across components
+Since network is shared in the same pod, endpoint ports should be unique across components, two components cannot have the same target port but two endpoints in a single component can have the same target port. only exception: container component with `dedicatedpod=true`
+
+### Commands:
+1. Id must be unique
+2. composite command:
+    - Should not reference itself via a subcommand
+    - Should not indirectly reference itself via a subcommand which is a composite command
+    - Should reference a valid devfile command
+    - Subcommands should be valid as well
+3. exec command should:
+map to a valid container component
+4. vscodeLaunch & vscodeTask: uri needs to be in valid uri form 
+5. `{build, run, test, debug}`, each kind of group can only have one default command associated with it.  If there are multiple commands of the same kind without a default, a warning will be displayed.
+
+
+### Components:
+1.  name must be unique
+2.  Container component:
+the container components must reference a valid volume component if it uses volume mounts, and the volume components are unique
```

Comment:
According to the markdown rules. These subitems of 2. are going to be displayed in one line without any punctuation mark.
![Screenshot_20201207_093335](https://user-images.githubusercontent.com/5887312/101322342-89d69980-386f-11eb-94c0-d463f7275257.png)

Please consider rewriting it in another way, like the following:
```markdown
2. Container component:
- the container components must reference a valid volume component if it uses volume mounts, and the volume components are unique;
- PROJECT_SOURCE or PROJECTS_ROOT are reserved Env, cannot be defined again in Env;
```

Maybe it even makes sense not to merge all components types in one list, and have something like

```markdown

### Components:
Common rules for all components types:
1. Name must be unique

#### Container component 
1. the container components must reference a valid volume component if it uses volume mounts, and the volume components are unique
2. PROJECT_SOURCE or PROJECTS_ROOT are reserved Env, cannot be defined again in Env

#### Plugin Component
1. Commands in plugins components share the same commands validation rules as listed above. Validation occurs after overriding and merging, in flattened devfile

#### Kubernetes & Openshift component 
1. URI needs to be in valid URI format

```